sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,290 @@
+	batch.sweeps[swapHash2] = *s
+
+	// Estimate weight of a batch with two sweeps.
+	batch2, err := estimateBatchWeight(batch)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create entity for sweep.
+	sweepEntity := &entity{
+		FeeRate:        s.minFeeRate,
+		NonCoopHint:    s.nonCoopHint,
+		IsExternalAddr: s.isExternalAddr,
+
+		// Calculate sweep weight as a difference.
+		CoopWeight:    batch2.CoopWeight - newBatch.CoopWeight,
+		NonCoopWeight: batch2.NonCoopWeight - newBatch.NonCoopWeight,
+	}
+
+	return sweepEntity, newBatch, nil
+}
+
+// estimateBatchWeight estimates batch weight.
+func estimateBatchWeight(batch *batch) (*entity, error) {
+	// Make sure the batch is not empty.
+	if len(batch.sweeps) == 0 {
+		return nil, errors.New("empty batch")
+	}
+
+	// Make sure fee rate is valid.
+	if batch.rbfCache.FeeRate < chainfee.AbsoluteFeePerKwFloor {
+		return nil, fmt.Errorf("feeRate is too low: %v",
+			batch.rbfCache.FeeRate)
+	}
+
+	// Find if the batch has at least one non-cooperative sweep.
+	hasNonCoop := false
+	for _, sweep := range batch.sweeps {
+		if sweep.nonCoopHint {
+			hasNonCoop = true
+		}
+	}
+
+	// Find some sweep of the batch. It is used if there is just one sweep.
+	var theSweep sweep
+	for _, sweep := range batch.sweeps {
+		theSweep = sweep
+		break
+	}
+
+	// Find sweep destination address (type) for weight estimations.
+	var destAddr btcutil.Address
+	if theSweep.isExternalAddr {
+		if theSweep.destAddr == nil {
+			return nil, errors.New("isExternalAddr=true, but " +
+				"destAddr is nil")
+		}
+		destAddr = theSweep.destAddr
+	} else {
+		// Assume it is taproot by default.
+		destAddr = (*btcutil.AddressTaproot)(nil)
+	}
+
+	// Make two estimators: for coop and non-coop cases.
+	var coopWeight, nonCoopWeight input.TxWeightEstimator
+
+	// Add output weight to the estimator.
+	err := sweeppkg.AddOutputEstimate(&coopWeight, destAddr)
+	if err != nil {
+		return nil, fmt.Errorf("sweep.AddOutputEstimate: %w", err)
+	}
+	err = sweeppkg.AddOutputEstimate(&nonCoopWeight, destAddr)
+	if err != nil {
+		return nil, fmt.Errorf("sweep.AddOutputEstimate: %w", err)
+	}
+
+	// Add inputs.
+	for _, sweep := range batch.sweeps {
+		// TODO: it should be txscript.SigHashDefault.
+		coopWeight.AddTaprootKeySpendInput(txscript.SigHashAll)
+
+		err = sweep.htlcSuccessEstimator(&nonCoopWeight)
+		if err != nil {
+			return nil, fmt.Errorf("sweep.htlcSuccessEstimator "+
+				"failed: %w", err)
+		}
+	}
+
+	return &entity{
+		BatchId:        batch.id,
+		FeeRate:        batch.rbfCache.FeeRate,
+		CoopWeight:     coopWeight.Weight(),
+		NonCoopWeight:  nonCoopWeight.Weight(),
+		NonCoopHint:    hasNonCoop,
+		IsExternalAddr: theSweep.isExternalAddr,
+	}, nil
+}
+
+// specialId is the value that indicates a new batch. It is returned by
+// selectBatch if the most cost-efficient action is new batch creation.
+const specialId = -1
+
+// entity is either batch of sweep and it holds data important for selection.
+type entity struct {
```

Comment:
I think `entity` is not necessarily the best name as it sounds a bit too general. Maybe consider something like `sweepSet` or `sweepEntity`? Also given that the role of this type is to store weight and fee information perhaps we can be even more specific: `sweepFeeDetails`?

Reply:
I renamed `entity` to `feeDetails`.